core/app/models/spree/user_address.rb

```diff
@@ -4,12 +4,18 @@ class UserAddress < Spree::Base
     belongs_to :address, class_name: "Spree::Address"
 
     validates_uniqueness_of :address_id, scope: :user_id
-    validates_uniqueness_of :user_id, conditions: -> { where(default: true) }, message: :default_address_exists, if: :default?
+    validates_uniqueness_of :user_id, conditions: -> { where(default: true, archived: false) }, message: :default_address_exists, if: :default?
```

Comment:
This conditions could be moved to 2 scopes: `default` and `not_archived`. This could be reusable in other places.